src/dmd/expressionsem.d

```diff
@@ -11343,13 +11344,48 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             }
             if (e.op == EXP.call)
             {
+
+                if (t1.ty == Tclass && t2.ty == Tclass)
```

Comment:
I think you also have to verify that the call is actually to Object.opCmp, as it can be redirected to other functions aswell, e.g.
```
class A
{
    int x;
    this(int a) { x = a; }
    
    alias opCmp = Object.opCmp;
    alias opCmp = my_cmp;
    
    final int my_cmp(A a)
    {
        return x - a.x;
    }
}

int main()
{
    auto a = new A(1);    
    return a < a;
} 
```
The comparison currently results in a non-virtual call to my_cmp. Does the lowered template do the same?

Reply:
The lowered template doest the same. I don't think the lowering should change the current behaviour.
Why do you want the behaviour to be different?

Reply:
> Why do you want the behaviour to be different?

I don't want observable changes, and AFAICT the actual call stays the same - that's good. But the template adds some additional checks that are not there so far. This might duplicate `null` checks in existing custom implementations of `opCmp`, or it might work differently. The spec doesn't seem to cover the comparison to `null`. I'm not against adding the checks, just saying they should be added to the spec (similar to `opEquals`) if the change is deliberate.

BTW: dmd is unable to inline the `__cmp` template, it can if it is written as ` return lhs is rhs ? 0 : lhs is null ? -1 : rhs is null ? 1 : lhs.opCmp(rhs);`. dmd doesn't eliminate unnecessary checks, though.

Reply:
> > Why do you want the behaviour to be different?
> 
> I don't want observable changes, and AFAICT the actual call stays the same - that's good. But the template adds some additional checks that are not there so far. This might duplicate `null` checks in existing custom implementations of `opCmp`, or it might work differently. The spec doesn't seem to cover the comparison to `null`. I'm not against adding the checks, just saying they should be added to the spec (similar to `opEquals`) if the change is deliberate.

The change is deliberate. I’ll update the spec to mention the changes.

> BTW: dmd is unable to inline the `__cmp` template, it can if it is written as ` return lhs is rhs ? 0 : lhs is null ? -1 : rhs is null ? 1 : lhs.opCmp(rhs);`. dmd doesn't eliminate unnecessary checks, though.

Hmm, rewritting it this way would make the code much harder to understand. Imho, we shouldn’t make the code harder to understand just so we improve dmd’s inlining. If this also affects ldc and gdc (haven’t checked yet) then we should modify the call.

I think the issue at hand is related to how dmd is able to inline such constructs. I think that should be a separate improvement done in dmd, as it will benefit other existing user code bases.